File: frontend/src/apis/customer/SetDefaultCustomerAddressApi.ts
import type { CustomerId, DeliveryAppState, SetDefaultCustomerAddressRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const setDefaultCustomerAddressApiDefinition = defineJsonPostApi1<CustomerId, SetDefaultCustomerAddressRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('customers')],
    [routeSegment('addresses'), routeSegment('default')],
  )

export function setDefaultCustomerAddress(
  customerId: CustomerId,
  payload: SetDefaultCustomerAddressRequest,
) {
  return postNormalizedDeliveryState(buildApiPath1(setDefaultCustomerAddressApiDefinition, customerId), payload)
}

File: backend/src/main/scala/services/customer/api/SetDefaultCustomerAddressApi.scala
package services.customer.api

import domain.shared.given

import cats.effect.IO
import services.customer.utils.setDefaultCustomerAddress
import domain.customer.SetDefaultCustomerAddressRequest
import domain.shared.{CustomerId, DeliveryAppState, UserRole}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import system.api.*
import system.app.*

val setDefaultCustomerAddressApi: FixedMethodApi1[CustomerId, DeliveryAppState] =
  jsonPostApi1[CustomerId, SetDefaultCustomerAddressRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("customers")),
    List(routeSegment("addresses"), routeSegment("default")),
  )
